src/main.rs

```diff
@@ -1,3 +1,201 @@
+    jobs: usize
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Person<'a>(Cow<'a, str>);
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Token<'a>(Cow<'a, str>);
+
+impl<'a> From<&'a str> for Token<'a> {
+    fn from(value: &'a str) -> Self {
+        Token(value.into())
+    }
+}
+impl<'a> From<String> for Token<'a> {
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
+    messages: Vec<Message<'a>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum ChatType {
+    PublicChannel,
+    PrivateChannel,
+    PublicSupergroup,
+    PrivateSupergroup,
+    PersonalChat,
+    ChatForbidden,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Message<'a> {
+    id: u64,
+    #[serde(rename = "type")]
+    message_type: MessageType,
+    date: NaiveDateTime,
+    from: Option<Person<'a>>,
+    text_entities: Vec<TextEntity>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum MessageType {
+    Service,
+    Message,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TextEntity {
+    #[serde(rename = "type")]
+    _type: TextEntityType,
+    text: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+enum TextEntityType {
+    Pre,
+    Bold,
+    Link,
+    Code,
+    Email,
+    Plain,
+    Phone,
+    Italic,
+    Cashtag,
+    Spoiler,
+    Mention,
+    Hashtag,
+    TextLink,
+    Underline,
+    BotCommand,
+    CustomEmoji,
+    MentionName,
+    Strikethrough,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ChatStatistics<'a> {
+    num_tokens: usize,
+    #[serde(borrow)]
+    members_tokens_map: HashMap<Person<'a>, HashMap<Token<'a>, usize>>,
+    #[serde(borrow)]
+    tokens_map: HashMap<Token<'a>, usize>,
+}
+impl<'a> ChatStatistics<'a> {
+    fn gather(chat: &'a Chat, jobs: usize) -> Self {
+        let iter = chat.messages.iter().filter(|message| message.message_type != MessageType::Service);
+
+        let num_messages = iter.clone().count();
+        let messages_per_thread = num_messages / jobs;
+        let (sender, receiver) = std::sync::mpsc::sync_channel(jobs);
+
+        let mut tokens_map = HashMap::new();
+        let mut members_tokens_map = HashMap::new();
+
+        std::thread::scope(|s| {
+            for i in 0..jobs {
+                let iter = iter.clone();
+                let sender = sender.clone();
+                s.spawn(move || {
+                    let mut chunk_tokens_map = HashMap::new();
+                    let mut chunk_members_tokens_map: HashMap<Person<'_>, HashMap<Token<'_>, usize>> = HashMap::new();
+
+                    for message in iter.skip(i * messages_per_thread).take(messages_per_thread) {
```

Comment:
This is very cool, and I appreciate the low-level approach.
However, I could suggest using [`rayon`](https://docs.rs/rayon/latest/rayon/index.html) crate in combination with `.chunks()` for parallelization instead of implementing your own. Take a look at the `.for_each` method:

```rust
iter.collect().chunks(messages_per_thread).collect().into_par_iter().for_each(|chunk| ...)
```

You can configure number of threads used by rayon using [ThreadPoolBuilder](https://docs.rs/rayon/latest/rayon/struct.ThreadPoolBuilder.html):

```rust
rayon::ThreadPoolBuilder::new().num_threads(22).build_global().unwrap();
```

(types may be off here and there)

Your method works just as well and is already implemented, but if you should decide to refactor your code (10-level pyramid of indentation is a lot), consider giving `rayon` a try.

Reply:
Thanks!
I've looked at rayon at first, but dropped it because thought there is no way to implement "chunking" (program used to process only one message in a thread, it was so slow that a a sequential processing was a way faster).
 
Thanks again, I'd definitely give it a go!